Blog about new implicit syntax in Scala 3 (#1168)
scala/scala-lang: _posts/2020-10-21-explicit-term-inference-in-scala-3.md

---
layout: blog-detail
post-type: blog
by: Meriam Lachkar, Vincenzo Bazzucchi, Scala Center
title: Explicit term inference with Scala 3
---

# Explicit term inference with Scala 3

One of the most striking changes for developers adopting Scala 3 is the introduction

> of new syntaxes to replace the implicit mechanism used in previous Scala versions.

Reviewer: "new syntaxes and semantics", perhaps?

Reviewer 2: Or just "new syntax"? I wouldn't normally _count_ syntax(es).

The motivation behind the new syntax is that the same `implicit` keyword was
used for different purposes and patterns and thus it became a way to express *how* to implement

> patterns. This means that when encountering this dreaded incantation, users need to decipher what

Reviewer: "dreaded" seems a bit dramatic, and a bit harsh on Scala 2 IMO. perhaps "shape-shifting" or "ambiguous" or something?

Scala 2 implicits weren't bad, it's just that we learned from the experience and now we know how to do even better.

the intent of the developer was: is this a conversion ? Does this avoid parameter repetition ?
Is this an extension of a type ? Is this a typeclass ? How do I import this ?

Seeing how pervasive implicits became in libraries and projects, Scala 3 aims at reducing confusion and cognitive load by using new keywords that convey the intent of the developer.

This post briefly introduces the new syntaxes available to Scala 3 programmers by analysing the

> most common use cases and patterns.

Reviewer: I'd suggest listing those syntaxes here (extension methods, ...), so the reader knows what's coming, and so they'll still get the basic idea even if they don't read the entire blog post. I always suggest this on every blog post: try to make the first few paragraphs a self-contained version of the entire post, so that if somebody only reads the start, they'll still get the important points.

---
layout: blog-detail
post-type: blog
by: Meriam Lachkar, Vincenzo Bazzucchi, Scala Center
title: Explicit term inference with Scala 3
---

# Explicit term inference with Scala 3

One of the most striking changes for developers adopting Scala 3 is the introduction
of new syntaxes to replace the implicit mechanism used in previous Scala versions.

The motivation behind the new syntax is that the same `implicit` keyword was
used for different purposes and patterns and thus it became a way to express *how* to implement
patterns. This means that when encountering this ambiguous incantation, users need to decipher what
the intent of the developer was: is this a conversion ? Does this avoid parameter repetition ?

> Is this an extension of a type ? Is this a typeclass ? How do I import this ?

Reviewer 2: Conventionally, there would be no space before an `?` in _English_ (unlike French).

Seeing how pervasive implicits became in libraries and projects, Scala 3 aims at reducing confusion and cognitive load by using new keywords that convey the intent of the developer.

This post briefly introduces the new syntaxes and semantics available to Scala 3 programmers by analysing the
most common use cases and patterns: extension methods, implicit parameters, implicit conversions and typeclasses.

## Compatibility disclaimer

While we believe that the new syntax represents an improvement, it is very important to highlight
that older code using `implicit` is still perfectly valid for the Scala 3.0 compiler, even if it will be deprecated in future releases.
You do not need to port your codebase right away, it can be an incremental and gradual process.

## Extension methods

When you do not have control over a type but you need to extend its capabilities with a new method,

> Scala 3 allows you to define an *extension method*

Reviewer 3: ```suggestion
Scala 3 allows you to define an *extension method*.
```

Assume that you are working with `List[Try[String]]` and that you often need to retrieve
the elements for which the computation succeeded.

Then you can extend this type to have a `collectSucceded` method:

```scala
// ListTryOps.scala

import scala.util.{Try, Success}

extension (ls: List[Try[String]]) def collectSucceeded: List[String] =
  ls.collect { case Success(x) => x }
```

> To remember the syntax, remark how `collectSucceeded` follows the object on which it will be available: `ls.collectSucceeded`.

Reviewer 3: ```suggestion
To remember the syntax, notice how `collectSucceeded` follows the object on which it will be available: `ls.collectSucceeded`.
```

> Note that extension methods can have type parameters as well:

Reviewer 3: ```suggestion
Extension methods can have type parameters as well:
```
Usually, we wouldn't use "note" to introduce something new. We only use that if it was deductible from what was said before, and we're only bringing attention to it.

```scala
extension[A] (ls: List[Try[A]]) def collectSucceeded: List[A] =
  ls.collect { case Success(x) => x }
```

Finally, you can add several methods without repeating the `extension` declaration:

```scala
extension[A] (ls: List[Try[A]]):
  def collectSucceeded: List[A] =
    ls.collect { case Success(x) => x }
  def getIndexOfFirstFailure: Option[Int] =
    ls.zipWithIndex.find((t, _) => t.isFailure)
      .map(_._2)
```

Extensions can be imported by name or using the `_` wildcard:

```scala
// Main.scala
import scala.util.Try
// import ListTryOps._ // Using wildcard
import ListTryOps.collectSucceeded

def getLastTweet(username: String): Try[String] = ???

@main def main =
  val niceTweets = List(getLastTweet("scala_lang"), getLastTweet("odersky")).collectSucceeded
  println(niceTweets)
```

### How was this done in Scala 2?

This pattern was particularly cumbersome to implement prior to Scala 3.
A typical approach would be the following:

```scala
implicit class ListTryExtension[A](val in: List[Try[A]]) extends AnyVal {
  def collectSucceeded: List[A] = ls.collect { case Success(a) => a }
  def getIndexOfFirstFailure: Option[Int] =

>       in
>         .zipWithIndex.find { case (t, _) => t.isFailure }
>         .map { case (_, index) => index }

Reviewer 3: ```suggestion
    in
      .zipWithIndex.find { case (t, _) => t.isFailure }
      .map { case (_, index) => index }
```
or even
```suggestion
    in.zipWithIndex.find { case (t, _) => t.isFailure }
      .map { case (_, index) => index }
```

}
```

> Remark that you need to define a name for the class, even if this is probably never going to be used besides the import statement.

Reviewer 3: ```suggestion
Note that you need to define a name for the class, even if this is probably never going to be used besides the import statement.
```

You also need to understand what `AnyVal` is, why it is good practice to extend it and
what its limitations are.

Some experience is required when reading this code to understand that its only goal is to
add a couple of methods to `List[Try[A]]`.

### Find out more

You can find more information about extension methods on [the dedicated documentation page](http://dotty.epfl.ch/docs/reference/contextual/extension-methods.html).

> I also suggest that you read how they complement another new Scala 3 feature: [opaque types](https://dotty.epfl.ch/docs/reference/other-new-features/opaques.html).

Reviewer 2: `We` is used everywhere except here and a couple of other places (I'll mark them), so maybe replace `I` with `we`.

    // ..
    def apply[T](body: => T)(using executor: ExecutionContext): Future[T] = // ...
}
```
We see here the other half of the syntax: if we declare a parameter as `using`, the compiler
will search the current scope at call sites for `given` values with a compatible type.

Note that the `executor` identifier is never used, so we can omit it:

```scala
given ExecutionContext = ExecutionContext.global
```
the same is possible for the `using` parameter:
```scala
def apply[T](body: => T)(using ExecutionContext): Future[T] = // ...
```
Note that in Scala 3 the type of a `given` definition must be explicit.

Another important aspect is how we import these values. For example, if you had a
more involved definition for the `ExecutionContext` that is used in multiple
files, you could refactor it into a different file:

```scala
// Context.scala
import scala.concurrent.ExecutionContext
import java.util.concurrent.Executors

object Context:
  given ExecutionContext =
    ExecutionContext.fromExecutor(Executors.newFixedThreadPool(5))
```

Then you can import it using a wildcard:

```scala
import Context.given
```
Or by making the type that you are bringing in scope explicit:
```

> Import Context.{given ExecutionContext}

Reviewer 2: Should be `import`, not `Import`.

```
This allows you to have more control over imports without relying on instance names.

### Find out more

You can learn more about `using` / `given` and about the rules of resolution in
[the documentation](https://dotty.epfl.ch/docs/reference/contextual/motivation.html).

### How was this done in Scala 2?

In Scala 2 this pattern was achieved by marking both the value and the parameter with the
`implicit` keyword. The previous example would look like this:

```scala
// Main.scala
import scala.concurrent._
object Main extends App
  implicit val ec: ExecutionContext = ExecutionContext.global
  Future(println("Hello World"))
}

// scala.concurrent.Future.scala
object Future {
    // ..
    def apply[T](body: => T)(implicit executor: ExecutionContext): Future[T] = // ...
}
```

We note again that we had to provide a name which might not be used for the variable. The `implictly` function from Scala 2 is renamed to `summon` in Scala 3.

> Finally the special import syntax allows users to explicitly import `given` instances (either by wildcard or by type) and not to have undesired instances obtained by a vague `import Context._`

Reviewer 3: The stronger added value of `import given` is that we can import by type. `import Context._` might be vague, but so is `import Foo.given`, and yet that still exists. I would put more emphasis on the type aspect, for example:

> Finally, the special import syntax allows users to import `given` instances by type, rather than by name. This makes more sense since we usually refer to them by type as well.

Author: Your formulation is clearer than mine, thanks

```
Use of implicit conversion class given_Conversion_Optional_Option in object OptionalConversion should be enabled
by adding the import clause 'import scala.language.implicitConversions'
or by setting the compiler option -language:implicitConversions.
See the Scala docs for value scala.language.implicitConversions for a discussion
why the feature should be explicitly enabled.
```
The warning can be silenced by adding `import scala.language.implicitConversions`, as
the message suggests.

### Find out more

You can learn more about this feature in the [documentation](https://dotty.epfl.ch/docs/reference/contextual/conversions.html).

### How was this done in Scala 2?

Scala 2 relied on `implicit` defs and `implicit` function values to implement this pattern:

```scala
import java.util.Optional

object OptionalConversion {
  implicit def optionalToOption[A](in: Optional[A]): Option[A] =
    if (in.isPresent) Some(in.get())
    else None
}
```

which could then be imported:
```scala
import OptionalConversion._
```

Note that at a glance, it is not clear that the definition is intended as an automatic conversion to an expected type:

>  - the name of the method can be a hint, but this is merely a convention that
>  other developers might not share
>  - the implicit def may be intended give you extension methods on its parameter type, depending on its result
>  - there is not straightforward way to verify that this function will not be used as implicit
>  argument for a function with an implicit parameter list

Reviewer 3: ```suggestion
 - the name of the method can be a hint, but this is merely a convention that
   other developers might not share
 - the implicit def may be intended give you extension methods on its parameter type, depending on its result
 - there is not straightforward way to verify that this function will not be used as implicit
   argument for a function with an implicit parameter list
```
otherwise I'm not sure it will properly render

The `Conversion` trait makes the definition explicit and more readable.

## A common pattern revisited in Scala 3: Typeclasses

This pattern is fundamental to Scala functional programming libraries such as [Cats](https://typelevel.org/cats/). In Scala 2 we used to heavily rely on `implicit` conversions to add methods
and on `implicit` parameters to propagate instances, which was a bit cryptic to beginners who were maybe already struggling with new concepts in functional programming.

In Scala 3, this pattern becomes simpler thanks to the new syntax.
Let's consider a simple typeclass such as `Show` which describes the capability of types to have a `String` representation.

We first describe the interface that all instances of this typeclass should implement:

```scala
// Show.scala
trait Show[A]:
  extension (a: A) def show: String
```
We can then add a companion object to provide a couple of auxiliary methods and instances
that make instance creation less tedious:
```scala
// Show.scala
object Show:
  def from[A](f: A => String): Show[A] = new Show[A]:
    extension (a: A) def show: String = f(a)

>   given[A: Show] as Show[List[A]] = new Show[List[A]]:

Reviewer 3: ```suggestion
  given[A: Show] as Show[List[A]]:
```

    extension (ls: List[A]) def show: String = ls.map(_.show).mkString(",")
```

To use it we need to import the interface and define an instance:
```scala
// Main.scala
import Show.{_, given}

case class Mountain(name: String, height: Int)

given Show[Mountain] =

>   Show.from((m: Mountain) => s"The ${m.name} is ${m.height} meters high")

Reviewer 2: This results in `The Mont Blanc is...` which isn't quite right, grammatically, even though `The Matterhorn is...` would be. Maybe just remove `The` and pick a different mountain? ;)

Reviewer 3: Hum, why is that? In French I'd say "Le Mont Blanc ...". Isn't that the same in English?

Author: Maybe I can just drop `The`

Reviewer 2: No... for reasons I can't explain. "I'm going to climb Mont Blanc" would be how we'd phrase it. Though if we had a white mountain with an English name, I'm sure we'd say "I'm going to climb _the_ White Mountain". But we also say "_the_ Matterhorn".

```scala
// Show.scala
trait Show[A] {
  def show(a: A): String
}
```
The second step was to define the extension method which relies on two implicits:
```scala
// ShowOps.scala
object ShowOps {
  implicit class showOps[A](in: A) extends AnyVal {
    def show(implicit instance: Show[A]): String =
      instance.show(in)
  }
}
```
And finally we can define the auxiliary methods in the companion object:
```scala
// Show.scala
object Show {
  def from[A](f: A => String): Show[A] = new Show[A] {
    def show(a: A): String = f(a)
  }

  // Either this or import ShowOps and use context bound
  implicit def showList[A](ls: List[A])(implicit instance: Show[A]) =
    ls.map(instance.show).mkList(",")
}
```
which brings us to the main definition:
```scala
// Main.scala
import Show._
import ShowOps._

case class Mountain(name: String, height: Int)

object Main extends App {
  implicit val mountainShow: Show[Mountain] =
    Show.fromFunction((m: Mountain) => s"The ${m.name} is ${m.height} meters high")

  val mountains = List(Mountain("Mont Blanc", 4808), Mountain("Matterhon", 4478))
  println(mountains.show)
}
```

> I believe that this example shows how `implicit` was used to achieve different goals

Reviewer 2: `I` -> `we`

> I believe that this example shows how `implicit` was used to achieve different goals
> and, in doing so, used to be more confusing:

Reviewer 3: Due to the first part of the sentence when "used to" is actually about *using* implicits, I read the second part of as "people used implicits to be more confusing" (i.e., people were doing their best to be confusing), and not as "implicits were more confusing back then". Consider:
```suggestion
I believe that this example shows how `implicit`s were used to achieve different goals
and, in doing so, how they used to be more confusing:
```
which creates the right symmetry around "how" rather than around "used to".

 - if you need to add methods, use `extension` rather than an implicit class
 - If you need an implicit parameters, use `using` to declare what you will need
 - If you are providing an implicit instance, use `given` to declare that the value is now available
 - If you need an implicit conversion, make it explicit for future readers
 - If you are never going to refer to a value by its name, do not provide one

### Find out more

Read more about typeclass implementation in Scala 3 in the [documentation](https://dotty.epfl.ch/docs/reference/contextual/type-classes.html).

Another extremely interesting, yet more advanced, feature in Scala 3 related to contextual abstractions are [Context Functions](https://dotty.epfl.ch/docs/reference/contextual/context-functions.html).

## Conclusion

We went over the main use cases of `implicit` in Scala 2 and offered a quick glance of what
they would look like in Scala 3. While the final result is almost the same, code is more
explicit and readable so that you can focus on solving your business problems rather than on
syntax.

> This is part of larger set of usability and ergonomy improvements for Scala 3 that we believe

Reviewer 3: ```suggestion
This is part of a larger set of usability and ergonomy improvements for Scala 3 that we believe
```
